### PR TITLE
[XPU] support speculate_limit_thinking_content_length for xpu

### PR DESCRIPTION
**Pull request by cmcamdy**
## Motivation

为 XPU 平台实现推测解码场景下的思考内容长度限制（thinking content length limiting）功能，
对齐 GPU 侧已有的 speculate_limit_thinking_content_length、limit_thinking_content_length 
和 reasoning_phase_token_constraint 算子能力。

## Modifications

1. 新增 XPU 算子：`limit_thinking_content_length`、`speculate_limit_thinking_content_length`、`reasoning_phase_token_constraint`，包含 kernel、wrapper、op 注册和 pybind 绑定。
2. 更新 `get_padding_offset` 接口，新增 `draft_tokens` 和 `seq_lens_encoder` 可选参数，返回值从 5 个减少为 4 个（移除 `cum_offsets_out`），并适配所有调用方。
3. 修复 `speculate_limit_thinking_content_length` kernel 中 `current_base_step` 计算逻辑和移除对 `step_idx` 的就地修改。
4. Python 层集成：在 `xpu_model_runner.py`、`xpu_pre_and_post_process.py`、`input_batch.py`、`sampler.py` 中接入新算子。
5. 新增单元测试：`test_limit_thinking_content_length.py`、`test_reasoning_phase_token_constraint.py`、`test_get_padding_offset.py`。

## Checklist

- [ ] Add at least a tag in the PR title.
  - Tag list: [`[FDConfig]`,`[APIServer]`,`[Engine]`, `[Scheduler]`, `[PD Disaggregation]`, `[Executor]`, `[Graph Optimization]`, `[Speculative Decoding]`, `[RL]`, `[Models]`, `[Quantization]`, `[Loader]`, `[OP]`, `[KVCache]`, `[DataProcessor]`, `[BugFix]`, `[Docs]`, `[CI]`, `[Optimization]`, `[Feature]`, `[Benchmark]`, `[Others]`, `[XPU]`, `[HPU]`, `[GCU]`, `[DCU]`, `[Iluvatar]`, `[Metax]`]
  - You can add new tags based on the PR content, but the semantics must be clear.
- [ ] Format your code, run `pre-commit` before commit.
- [ ] Add unit tests. Please write the reason in this PR if no unit tests.
- [ ] Provide accuracy results.
- [ ] If the current PR is submitting to the `release` branch, make sure the PR has been submitted to the `develop` branch, then cherry-pick it to the `release` branch with the `[Cherry-Pick]` PR tag.

### CODE REVIEW

File: custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/limit_thinking_content_length.xpu [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    const int64_t* inject_token_ids,
+    const int bs,
+    const int eos_token_id_len,
+    const int inject_len,
+    const bool splitwise_role_is_decode) {
+  int cid = core_id();
+  int ncores = core_num();
+  int clusterid = cluster_id();
+  int nclusters = cluster_num();
+  if (clusterid != 0) return;
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/plugin/src/kernel/kunlun3cpp/limit_thinking_content_length.xpu:54]: 已添加

**Comment by cmcamdy** [comment]
> > 🤖 AI Code Review | `2026-04-20 18:52:06`
> 
> ## 📋 Review 摘要
> **PR 概述**：为 XPU 平台新增 `speculate_limit_thinking_content_length`、`limit_thinking_content_length`、`reasoning_phase_token_constraint` 三个算子的完整实现（kernel + wrapper + op 注册 + pybind），并重构 `get_padding_offset` 接口以对齐 GPU 侧签名。 **变更范围**：`custom_ops/xpu_ops/`（C++ kernel/wrapper/op）、`fastdeploy/` Python 调用层 **影响面 Tag**：`XPU` `OP` `Speculative Decoding`
> 
> ### 📝 PR 规范检查
> PR 描述中 **Motivation** 和 **Modifications** 均未填写。
> 
> **描述建议**（可直接复制）：
> 
> ```
> ## Motivation
> 
> 为 XPU 平台补齐推测解码场景下 thinking content 长度限制相关算子，使 XPU 与 GPU 对齐。
> 
> ## Modifications
> 
> 1. 新增 XPU 算子：limit_thinking_content_length、speculate_limit_thinking_content_length、reasoning_phase_token_constraint
> 2. 重构 get_padding_offset 接口对齐 GPU 侧签名
> 3. 修复 speculate_limit_thinking_content_length kernel 中 current_base_step 计算逻辑
> 4. Python 层适配及新增单测
> ```
> 
> ### 问题
> 级别	文件	概述
> 🔴 Bug	`speculate_limit_thinking_content_length.cpp`	CPU wrapper 的 `current_base_step` 未与 XPU kernel 同步更新
> ❓ 疑问	`get_padding_offset.cc`	draft_tokens 传入时直接 PD_THROW，错误消息未说明 XPU 平台限制
> ### 🔴 Bug 详情：CPU wrapper 与 XPU kernel 的 `current_base_step` 逻辑不一致
> **文件**：`custom_ops/xpu_ops/src/plugin/src/wrapper/mtp_wrapper/speculate_limit_thinking_content_length.cpp` 第 77 行
> 
> XPU kernel（`.xpu` 文件）已将 `current_base_step` 改为 `step_idx[bid] + 1`，并移除了末尾的 `step_idx -= discarded_tokens` 回退逻辑（与 GPU 侧 `.cu` 对齐）。但 CPU wrapper 仍使用旧逻辑：
> 
> ```c++
> // CPU wrapper（当前代码，第 77 行）—— 旧逻辑
> const int64_t current_base_step = step_idx[bid] - original_accept_num + 1;
> // ...
> // 第 157-159 行 —— 已被 kernel 移除的回退逻辑
> const int discarded_tokens = original_accept_num - new_accept_num;
> if (discarded_tokens > 0) {
>     step_idx[bid] -= discarded_tokens;
> }
> ```
> 
> CPU 路径用于单测和调试，逻辑不一致会导致 CPU/XPU 两条路径行为不同，单测无法准确验证 kernel 正确性。
> 
> **建议**：将 CPU wrapper 第 77 行同步修改为 `step_idx[bid] + 1`，并移除 157-159 行的回退代码。
> 
> ### ❓ 疑问：`get_padding_offset.cc` 中 draft_tokens 的 PD_THROW
> **文件**：`custom_ops/xpu_ops/src/ops/get_padding_offset.cc` 第 55 行
> 
> 签名新增 `draft_tokens` 和 `seq_lens_encoder` 可选参数以对齐 GPU 侧，但传入非空值时直接抛异常。当前 Python 侧非投机路径传入 `None, None`，投机路径走 `speculate_pre_process`，运行时不会触发。建议在错误消息中注明这是 XPU 平台的已知限制。
> 
> ### 总体评价
> 整体实现完整，kernel/wrapper/op/pybind/Python 各层配套齐全，测试覆盖好。建议修复 CPU wrapper 与 kernel 的逻辑不一致后合入。

已修复

File: custom_ops/xpu_ops/src/ops/get_padding_offset.cc [hunk 2]
```diff
@@ -40,49 +52,53 @@ std::vector<paddle::Tensor> GetPaddingOffset(const paddle::Tensor& input_ids,
       paddle::full({bsz + 1}, 0, paddle::DataType::INT32, input_ids.place());
 
   if (token_num_data > 0) {
+    if (draft_tokens || seq_lens_encoder) {
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/get_padding_offset.cc:55]: 预埋接口，为后续speculate与nomal合并做准备，上层不会传入

File: custom_ops/xpu_ops/src/ops/get_padding_offset.cc [hunk 3]
```diff
@@ -16,22 +16,34 @@
...
+  const int max_seq_len = input_ids_shape[1];
   const int token_num_data = static_cast<int>(cpu_token_num);
 
   auto x_remove_padding = paddle::full(
-      {token_num_data}, 0, paddle::DataType::INT64, input_ids.place());
+      {token_num_data}, 2, paddle::DataType::INT64, input_ids.place());
   auto batch_id_per_token = paddle::full(
-      {token_num_data}, 0, paddle::DataType::INT32, input_ids.place());
+      {token_num_data}, -1, paddle::DataType::INT32, input_ids.place());
   auto cum_offsets_out =
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/get_padding_offset.cc]: 已修复

File: custom_ops/xpu_ops/src/ops/mtp/reasoning_phase_token_constraint.cc [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    const paddle::Tensor& cu_seqlens_q_output,
+    const paddle::Tensor& enable_thinking,
+    int64_t think_end_id,
+    int64_t line_break_id) {
+  phi::XPUPlace place(phi::backends::xpu::GetXPUCurrentDeviceId());
+  auto dev_ctx = paddle::experimental::DeviceContextPool::Instance().Get(place);
+  auto xpu_ctx = static_cast<const phi::XPUContext*>(dev_ctx);
+  baidu::xpu::api::Context* ctx = xpu_ctx->x_context();
+
+  if (logits.is_cpu()) {
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/mtp/reasoning_phase_token_constraint.cc:44]: 已修复

File: custom_ops/xpu_ops/src/ops/get_padding_offset.cc [hunk 5]
```diff
@@ -16,73 +16,86 @@
...
       paddle::full({bsz + 1}, 0, paddle::DataType::INT32, input_ids.place());
   auto cu_seqlens_k =
       paddle::full({bsz + 1}, 0, paddle::DataType::INT32, input_ids.place());
 
   if (token_num_data > 0) {
+    if (draft_tokens || seq_lens_encoder) {
+      // TODO(chenhuan09) : support speculate mode
+      PD_THROW("Draft tokens are not supported on XPU currently.");
+    }
     int r =
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/get_padding_offset.cc:59]: 已经修复

**Review by hong19860320 — DISMISSED** [review]
LGTM

File: custom_ops/xpu_ops/src/ops/get_padding_offset.cc [hunk 6]
```diff
@@ -16,73 +16,85 @@
...
   std::vector<int64_t> input_ids_shape = input_ids.shape();
   const int bsz = seq_len.shape()[0];
-  const int seq_length = input_ids_shape[1];
+  const int max_seq_len = input_ids_shape[1];
   const int token_num_data = static_cast<int>(cpu_token_num);
 
   auto x_remove_padding = paddle::full(
-      {token_num_data}, 0, paddle::DataType::INT64, input_ids.place());
+      {token_num_data}, 2, paddle::DataType::INT64, input_ids.place());
   auto batch_id_per_token = paddle::full(
```
[the remark this answers is not in the document]
Comment (cmcamdy) [reply custom_ops/xpu_ops/src/ops/get_padding_offset.cc:47]: 对齐GPU

**Review by hong19860320 — APPROVED** [review]
LGTM

**Review by Jiang-Jia-Jun — APPROVED** [review]

**Review by zyyzghb — APPROVED** [review]

**Review by yongqiangma — APPROVED** [review]
LGTM
